pkg/controller/daemon/daemoncontroller_test.go

```diff
@@ -471,6 +481,24 @@ func TestInsufficentCapacityNodeDaemonDoesNotUnscheduleRunningPod(t *testing.T)
 	syncAndValidateDaemonSets(t, manager, ds, podControl, 0, 0)
 }
 
+// DaemonSets should only place onto nodes with sufficient free resource and matched node selector
+func TestInsufficientCapacityNodeSufficientCapacityWithNodeLabelDaemonLaunchPod(t *testing.T) {
+	podSpec := resourcePodSpecWithoutNodeName("50M", "75m")
+	ds := newDaemonSet("foo")
+	ds.Spec.Template.Spec = podSpec
+	ds.Spec.Template.Spec.NodeSelector = simpleNodeLabel
+	manager, podControl, _ := newTestController(ds)
+	node1 := newNode("not-enough-resource", nil)
+	node1.Status.Allocatable = allocatableResources("10M", "20m")
+	node2 := newNode("enough-resource", simpleNodeLabel)
+	node2.Status.Allocatable = allocatableResources("100M", "200m")
+	manager.nodeStore.Add(node1)
+	manager.nodeStore.Add(node2)
+	manager.dsStore.Add(ds)
+	// TODO enable `syncAndValidateDaemonSets` can check event number
```

Comment:
Not sure why checking event number, suggest to add `eventRecorder` into `daemonSetsController`, so we can check whether unexpected events are triggered.

BTW, if updating to `syncAndValidateDaemonSets` is required, prefer to handle it in one PR, otherwise, we'll lose the context.